### PR TITLE
wshowkeys: init at 2020-03-29

### PR DESCRIPTION
**Pull request by primeos**
###### Motivation for this change

Only a draft for discoverability (not sure if that makes sense, but there where a few requests already and maybe someone would like to use it already).

I've tested it and it works but the software itself seems still alpha/beta quality and I don't think it's ready for Nixpkgs yet (it doesn't crash, but there are some rendering issues, the overlay doesn't disappear automatically after a few secondes, etc. and we should wait for a tagged (pre-)release anyway).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by berbiche** [comment]
@primeos Was this closed because wshowkeys seem unmaintained?

**Comment by primeos** [comment]
Kinda. It should still be maintained (especially if you send Drew a patch) but there isn't any active development anymore. However, while it's probably feature complete it doesn't really seem that usable/polished yet. And without a tagged release (which I was hoping for after a few more patches) it doesn't really make much sense to merge this into Nixpkgs.

But it seems like e.g. Fedora packages it already: https://repology.org/project/wshowkeys/versions

File: nixos/modules/programs/wshowkeys.nix [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  cfg = config.programs.wshowkeys;
+in {
+  meta.maintainers = with maintainers; [ primeos ];
+
+  options = {
+    programs.wshowkeys = {
+      enable = mkEnableOption ''
+        wshowkeys (displays keypresses on screen on supported Wayland
+        compositors). It requires root permissions to read input events, but
+        these permissions are dropped after startup'';
```
Comment (berbiche) [inline nixos/modules/programs/wshowkeys.nix:15]: Does it work without root if you are part of the `input` group?

**Review by berbiche — COMMENTED** [review]
I think this package is relevant to the wayland community and should be reopened.
I propose adding myself to the list of maintainers if you wish.

File: nixos/modules/programs/wshowkeys.nix (hunk 1 above)
Comment (primeos) [reply nixos/modules/programs/wshowkeys.nix:15]: No, doesn't look like it: https://git.sr.ht/~sircmpwn/wshowkeys/tree/6388a49e0f431d6d5fcbd152b8ae4fa8e87884ee/devmgr.c#L127

But looking at the rest of the code it should be enough to be in the `input` group. It should only require a small patch to detect this and skip the setuid related code.

**Comment by primeos** [comment]
> I think this package is relevant to the wayland community and should be reopened.

Yeah, I'm not aware of any alternatives and it can clearly be useful. But a tagged release would be helpful, though not required.
I'll reopen this PR for now but it's probably a good idea to quickly ask Drew first if it's ok to package this already.

> I propose adding myself to the list of maintainers if you wish.

That would be helpful as I didn't really need it myself so far.

**Comment by primeos** [comment]
> I'll reopen this PR for now but it's probably a good idea to quickly ask Drew first if it's ok to package this already.

Ok, I finally asked Drew via IRC and he's fine with it :)
So this is now ready for a final review and merging (if there aren't any objections).

File: pkgs/tools/wayland/wshowkeys/default.nix [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    description = "Displays keys being pressed on a Wayland session";
+    longDescription = ''
+      Displays keypresses on screen on supported Wayland compositors (requires
+      wlr_layer_shell_v1 support).
+      Note: This tool requires root permissions to read input events, but these
+      permissions are dropped after startup. The NixOS module provides such a
+      setuid binary (use "programs.wshowkeys.enable = true;").
+    '';
+    homepage = "https://git.sr.ht/~sircmpwn/wshowkeys";
+    license = licenses.mit;
```
Comment (ryantm) [inline pkgs/tools/wayland/wshowkeys/default.nix:31]: It looks like this thing is both gpl3Only and mit, mixed license.

**Comment by ryantm** [comment]
Result of `nixpkgs-review pr 83687` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
<br>- tests.nixos-functions.nixos-test
</details>
<details>
  <summary>1 package built:</summary>
<br>- wshowkeys
</details>

**Review by ryantm — APPROVED** [review]

File: pkgs/tools/wayland/wshowkeys/default.nix (hunk 2 above)
Comment (primeos) [reply pkgs/tools/wayland/wshowkeys/default.nix:31]: Good catch, I somehow messed that up, thanks!
I pushed a fix (8a1fdce8d3482b60ac13678a8eab838777b51549) and will ask Drew if it should be `gpl3Only` or `gpl3Plus` after my related Meson issue is resolved (whether to recommend SPDX license identifiers or not).

**Pull request merged**